source/cppfront.cpp

```diff
@@ -1664,6 +1664,66 @@ class cppfront
         }
     }
 
+    // Don't work yet, TODO: finalize deducing pointer types from parameter lists
+    auto is_pointer_declaration(parameter_declaration_list_node const* decl_node, int deref_cnt, int addr_cnt) -> bool {
+        return false;
+    }
+
+    auto is_pointer_declaration(declaration_node const* decl_node, int deref_cnt, int addr_cnt) -> bool {
+        if (!decl_node) return false;
+        if (addr_cnt > deref_cnt) return true;
+
+        return std::visit([&](auto const& type){
+            return is_pointer_declaration(type.get(), deref_cnt, addr_cnt);
+        }, decl_node->type);
+    }
```

Comment:
Can you elaborate on what the `std::visit` is doing?

Reply:
I found `std::visit()` a convenient way to extract the type from `decl_node->type` and pass a pointer to the appropriate overload.

My previous implementation consists of many `std::get_if()` with if statements that look like this (it is before moving to `type_id` node):
```cpp
// if it is a function check if it returns pointer type
if (auto* fun = std::get_if<declaration_node::function>(&decl->declaration->type)) {
    if (auto* ret_id_expr = std::get_if<function_type_node::id>(&(*fun)->returns)) {
        if (auto* unqual = std::get_if<id_expression_node::unqualified>(&(*ret_id_expr)->id)){
            return is_it_pointer_declaration(*unqual, deref_cnt, addr_cnt);
        }
    }
}

// if it is an object declaration check if we declare an pointer type
if (auto* obj_id_expr = std::get_if<declaration_node::object>(&decl->declaration->type)) {
    if (auto* unqual = std::get_if<id_expression_node::unqualified>(&(*obj_id_expr)->id)){
        return is_it_pointer_declaration(*unqual, deref_cnt, addr_cnt);
    }
}

// if we initialize an object with a function check if it is initialized with a pointer type
if (decl->initializer && decl->initializer->suspicious_initialization) {
    auto init_decl = sema.get_declaration_of(*decl->initializer->suspicious_initialization, true);
    if (init_decl && init_decl->declaration) {
        if (auto* fun = std::get_if<declaration_node::function>(&init_decl->declaration->type)) {
            if (auto* ret_id_expr = std::get_if<function_type_node::id>(&(*fun)->returns)) {
                if (auto* unqual = std::get_if<id_expression_node::unqualified>(&(*ret_id_expr)->id)){
                    return is_it_pointer_declaration(*unqual, deref_cnt, addr_cnt);
                }
            }
        }
    }
}
```

Reply:
I personally think that `std::visit` is what allows to fully unleash the power of `std::variant`. In particular the idiom presented at the end of [the cppreference example](https://en.cppreference.com/w/cpp/utility/variant/visit#Example) leaves all the type checks to the compiler (i.e. no explicit pointers checked for nullity) and provides very nice syntactic sugar.
In fact `inspect` from Cpp2 made me wonder about a twist of that idiom that would be even nicer syntactic sugar with a more left-to-right taste. [Here is a quick example](https://github.com/hsutter/cppfront/compare/main...jarzec:cppfront:inspect-in-cpp1) of a version of `inspect` for Cpp1 with an example application in cppfront that passes all regression tests. I know from experience that this kind of syntax for handling `variant`s is used easily even by less experienced C++ users.

Reply:
Cool, thanks!